### PR TITLE
Minutly Probing for Custom Metrics

### PR DESCRIPTION
**Pull request by jvanbaarsen**
This adds the first setup for the probing system in the Elixir side.

We use a Genserver that holds all the probes, this Genserver will schedule itself every minute, and then create a Task for every Probe.

The Probes Genserver is supervised by appsignal.ex.

@jeffkreeftmeijer I'm not 100% sure if this tests are enough coverage.

Next steps would be:
1. Create a way to allow probes to be their own Genservers (Some form of Init system)
2. Create the Erlang Probe

### CODE REVIEW

**Comment by jeffkreeftmeijer** [comment]
To make sure the probes work, we could register a probe in the test and make sure it's called. Maybe an agent, like the mocks in test/support, which counts how often a function on it is called and a probe that calls out to it?

**Comment by jvanbaarsen** [comment]
@jeffkreeftmeijer Interesting! I'll look into that

File: test/appsignal/probes/probes_test.exs [hunk 1]
```diff
@@ -0,0 +1,15 @@
+defmodule Appsignal.Probes.ProbesTest do
+  use ExUnit.Case, async: false
+
+  alias Appsignal.Probes
+
+  describe "register/2" do
+    test "Does register a probe when given a function as probe" do
```
Comment (tombruijn) [inline test/appsignal/probes/probes_test.exs]:
```suggestion
    test "does register a probe when given a function as probe" do
```

File: test/appsignal/probes/probes_test.exs [hunk 2]
```diff
@@ -0,0 +1,15 @@
...
+      assert :ok == Probes.register(:some_probe, fn -> nil end)
+    end
+
+    test "Returns an error tupple when probe is not a function" do
```
Comment (tombruijn) [inline test/appsignal/probes/probes_test.exs]:
```suggestion
    test "returns an error tupple when probe is not a function" do
```

File: lib/appsignal/probes/probes.ex [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+    {:noreply, probes}
+  end
+
+  defp genserver_running? do
+    pid = Process.whereis(__MODULE__)
+    !is_nil(pid) && Process.alive?(pid)
+  end
+
+  defp schedule_probes do
+    Process.send_after(self(), :run_probes, 60 * 1000)
```
Comment (tombruijn) [inline lib/appsignal/probes/probes.ex]: Should be scheduled every first second of the minute. If we want to follow the Ruby implementation.

Comment (jvanbaarsen) [reply lib/appsignal/probes/probes.ex]: 👍

**Comment by jvanbaarsen** [comment]
@jeffkreeftmeijer Can you do a re-review? I added an integration test for the probing the way we discussed.

File: lib/appsignal/probes/probes.ex [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+  end
+
+  defp genserver_running? do
+    pid = Process.whereis(__MODULE__)
+    !is_nil(pid) && Process.alive?(pid)
+  end
+
+  if Mix.env() == :test do
+    defp schedule_probes do
+      IO.puts "Schedule?"
```
Comment (jeffkreeftmeijer) [inline lib/appsignal/probes/probes.ex]:
```suggestion
```

File: lib/appsignal/probes/probes.ex [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+  end
+
+  def init([]) do
+    schedule_probes()
+    {:ok, %{}}
+  end
+
+  def handle_cast({name, probe}, probes) do
+    if Map.has_key?(probes, name) do
+      Logger.debug(fn -> "A probe with name '#{name}' already exists. Overriding that one" end)
```
Comment (jeffkreeftmeijer) [inline lib/appsignal/probes/probes.ex]:
```suggestion
      Logger.debug(fn -> "A probe with name '#{name}' already exists. Overriding that one." end)
```

File: test/appsignal/probes/probes_test.exs [hunk 6]
```diff
@@ -0,0 +1,38 @@
+defmodule Appsignal.Probes.ProbesTest do
+  use ExUnit.Case, async: false
+
+  alias Appsignal.Probes
+  alias FakeProbe
+
+  describe "register/2" do
+    test "does register a probe when given a function as probe" do
```
Comment (jeffkreeftmeijer) [inline test/appsignal/probes/probes_test.exs]:
```suggestion
    test "registers a probe when given a function as probe" do
```

**Comment by jvanbaarsen** [comment]
@jeffkreeftmeijer 	Thanks for reviewing, some of those things are from me actually debugging now. I'll update the PR :)

File: lib/appsignal/probes/probes.ex [hunk 7]
```diff
@@ -0,0 +1,84 @@
...
+
+    {:noreply, Map.put(probes, name, probe)}
+  end
+
+  def handle_cast(name, probes) do
+    {:noreply, Map.delete(probes, name)}
+  end
+
+  def handle_info(:run_probes, probes) do
+    Logger.debug(fn -> "Running #{Enum.count(probes)} probes" end)
```
Comment (jeffkreeftmeijer) [inline lib/appsignal/probes/probes.ex]:
```suggestion
    Logger.debug(fn -> "Running #{Enum.count(probes)} probe(s)" end)
```

**Comment by jeffkreeftmeijer** [comment]
While trying the Erlang probes in a Phoenix app with debug logging on, I noticed two debug messages pop up every minute:

```
[debug] Running 1 probes
[debug] Creating task for erlang
```

I'm not sure if it's useful to print both of these every minute, even with the debug level. Also, could we prefix these with "AppSignal: ", so it's clear where these come from?

**Comment by jvanbaarsen** [comment]
@jeffkreeftmeijer I mirrored the logging from the ruby gem. I think the biggest difference is that with the Ruby gem we don't "pollute" the console, and we send this to a different log file. I would be fine by removing the logging all together in this case. What do you say?

**Pull request merged**